sksparse/cholmod.pyx

```diff
@@ -287,77 +287,94 @@ cdef int _np_typenum_for_index(int itype):
-    assert m.itype == common._common.itype
+    # This is a little tricky: We build 3 arrays, views on each part of the
+    # cholmod_dense object. They all have the same _CholmodSparseDestructor
+    # object as base. So none of them will be deallocated until they have all
+    # become unused. Then those are built into a csc_matrix.
 
+    # init destructor for cholmod data
+    cdef _CholmodSparseDestructor base = _CholmodSparseDestructor()
+    base.init(m, common)
+
+    # convert to NumPy arrays
+    assert m.itype == common._common.itype
     cdef np.ndarray indptr = np.PyArray_SimpleNewFromData(
         1, [m.ncol + 1], _np_typenum_for_index(m.itype), m.p)
-    PyArray_ENABLEFLAGS(indptr, np.NPY_WRITEABLE)
-    np.set_array_base(indptr, cleaner)
     cdef np.ndarray indices = np.PyArray_SimpleNewFromData(
         1, [m.nzmax], _np_typenum_for_index(m.itype), m.i)
-    PyArray_ENABLEFLAGS(indices, np.NPY_WRITEABLE)
-    np.set_array_base(indices, cleaner)
     cdef np.ndarray data = np.PyArray_SimpleNewFromData(
         1, [m.nzmax], _np_typenum_for_data(m.xtype), m.x)
-    PyArray_ENABLEFLAGS(data, np.NPY_WRITEABLE)
-    np.set_array_base(data, cleaner)
 
+    # set destructor and check if writeable
+    for array in (indptr, indices, data):
+        np.set_array_base(array, base)
+        assert np.PyArray_ISWRITEABLE(indptr)
+
+    # return sparse matrix
     return sparse.csc_matrix((data, indices, indptr), shape=(m.nrow, m.ncol))
 
-cdef class _DenseCleanup:
-    cdef cholmod_dense * _dense
+cdef class _CholmodDenseDestructor:
+    """This is a destructor for NumPy arrays based on dense data of Cholmod.
+    Use this only once for each Cholmod dense array. Otherwise memory will be
+    freed multiple times."""
+    cdef cholmod_dense* _dense
     cdef Common _common
+
+    cdef init(self, cholmod_dense* m, Common common):
+        assert m is not NULL
+        assert common is not None
+        self._dense = m
+        self._common = common
+
     def __dealloc__(self):
         if self._common._use_long:
             cholmod_c_free_dense = cholmod_l_free_dense
         else:
             cholmod_c_free_dense = cholmod_free_dense
         cholmod_c_free_dense(&self._dense, &self._common._common)
 
-cdef _py_dense(cholmod_dense * m, Common common):
-    """Build an ndarray that's a view onto m, with a 'base' with appropriate
-    destructor. 'm' must have been allocated by cholmod."""
-
-    assert m is not NULL
-    assert common is not None
-    # Construct cleaner first, so even if we later raise an exception we still
-    # fulfill the contract that we will take care of cleanup:
-    cdef _DenseCleanup cleaner = _DenseCleanup()
-    cleaner._dense = m
-    cleaner._common = common
-
-    cdef np.ndarray out = np.PyArray_SimpleNewFromData(
-        1, [m.ncol * m.nrow], _np_typenum_for_data(m.xtype), m.x).reshape((m.ncol, m.nrow)).T
-    PyArray_ENABLEFLAGS(out, np.NPY_WRITEABLE)
-    np.set_array_base(out, cleaner)
```

Comment:
The bug was here or to be precise two lines above. Due to the `reshape` and `transpose` two lines above, the `out` array was not anymore the array constructed from the Cholmod data. This is the reason why the `base` attribute was already set.